File: solutions/typescript/090-find-minimum-in-rotated-sorted-array.ts
/**
 * Problem 90: Find Minimum in Rotated Sorted Array (LeetCode 153)
 * Difficulty: Med
 * Language: TypeScript
 */
function findMin(nums: number[]): number {
  const go = (lo: number, hi: number): number => {
    if (lo === hi) return nums[lo];
    const mid = Math.floor((lo + hi) / 2);
    return nums[mid] > nums[hi] ? go(mid + 1, hi) : go(lo, mid);
  };
  return go(0, nums.length - 1);
}

File: solutions/ocaml/090-find-minimum-in-rotated-sorted-array.ml
(*
  Problem 90: Find Minimum in Rotated Sorted Array (LeetCode 153)
  Difficulty: Med
  Language: OCaml
*)
let find_min nums =
  let a = Array.of_list nums in
  let rec go lo hi =
    if lo = hi then a.(lo)
    else let mid = (lo + hi) / 2 in
      if a.(mid) > a.(hi) then go (mid + 1) hi else go lo mid
  in go 0 (Array.length a - 1)
